src/net_processing.cpp

```diff
@@ -451,7 +511,305 @@ struct Peer {
+         * a capacity of the sketch sent out initially, so that a sketch extension is of the same size.
+         */
+        uint16_t m_capacity_snapshot{0};
+
+        /**
+         * In a reconciliation round initiated by us, if we asked for an extension, we want to store
+         * the sketch computed/transmitted in the initial step, so that we can use it when
+         * sketch extension arrives.
+         */
+        std::vector<uint8_t> m_remote_sketch_snapshot;
+
+        /**
+        * Reconciliation sketches are computed over short transaction IDs.
+        * Short IDs are salted with a link-specific constant value.
+        */
+        uint32_t ComputeShortID(const uint256 wtxid)
+        {
+            uint64_t k0 = m_salt.GetUint64(0);
+            uint64_t k1 = m_salt.GetUint64(1);
+            uint64_t s = SipHashUint256(k0, k1, wtxid);
+            uint32_t short_txid = 1 + (s & 0xFFFFFFFF);
+            return short_txid;
+        }
+
+        /**
+         * Estimate a capacity of a sketch we will send or use locally (to find set difference)
+         * based on the local set size.
+         */
+        uint16_t EstimateSketchCapacity() {
+            uint16_t set_size_diff = std::abs(uint16_t(m_local_set.size()) - m_remote_set_size);
+            uint16_t min_size = std::min(uint16_t(m_local_set.size()), m_remote_set_size);
+            uint16_t weighted_min_size = m_remote_q * min_size;
+            uint16_t estimated_diff = 1 + weighted_min_size + set_size_diff;
+            return minisketch_compute_capacity(RECON_FIELD_SIZE, estimated_diff, RECON_FALSE_POSITIVE_COEF);
+        }
+
+        /**
+         * Reconciliation involves computing a space-efficient representation of transaction identifiers (a sketch).
+         * A sketch has a capacity meaning it allows reconciling at most a certain number of elements. (see BIP-330).
+         * Considering whether we are going to send a sketch to a peer or use locally, we estimate the set difference.
+         */
+        minisketch* ComputeSketch(const std::set<uint256> local_set, uint16_t& capacity)
+        {
+            std::vector<uint32_t> short_ids;
+            for (uint256 wtxid: local_set) {
+                uint32_t short_txid = ComputeShortID(wtxid);
+                short_ids.push_back(short_txid);
+                m_local_short_id_mapping.insert(std::pair<uint32_t, uint256>(short_txid, wtxid));
+            }
+
+            capacity = std::min(capacity, MAX_SKETCH_CAPACITY);
+
+            if (short_ids.size() == 0) return nullptr;
+            minisketch* sketch = minisketch_create(RECON_FIELD_SIZE, 0, capacity);
+            if (sketch == nullptr) {
+                return nullptr;
+            }
+            for (uint32_t short_id: short_ids) {
+                minisketch_add_uint64(sketch, short_id);
+            }
+            return sketch;
+        }
+
+        minisketch* ComputeExtendedSketch() {
```

Comment:
I've PR'ed https://github.com/sipa/minisketch/pull/28 that adds a safer C++ wrapper around the minisketch* type. You may want to use it instead of doing manual memory management for them. It also adds some convenience functions for dealing with fpbits (CreateFP and DecodeFP).